File: 数据预处理/插值/code1.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.font_manager import FontProperties
import os
from scipy.interpolate import PchipInterpolator, CubicSpline

def get_chinese_font():
    """
    返回指定路径的微软雅黑字体 FontProperties 实例。
    """
    path = '/home/xhwang/msyh.ttc'
    if os.path.exists(path):
        return FontProperties(fname=path)
    else:
        # 字体文件不存在则回退到 matplotlib 默认字体（可能无法显示中文）
        return FontProperties()

plt.rcParams['axes.unicode_minus'] = False
chinese_font = get_chinese_font()

# 定义样本点
x = np.arange(1, 11)            # 1,2,...,10
y = np.log(x)                   # ln(x)

# 绘制样本点
plt.plot(x, y, 'o', label='样本点')

# 细分网格
new_x = np.arange(0.01, 10.01, 0.1)

# 三次埃尔米特插值（PCHIP）
pchip_interp = PchipInterpolator(x, y)
pchip_y = pchip_interp(new_x)
plt.plot(new_x, pchip_y, '-', label='三次埃尔米特插值')

# 三次样条插值
spline_interp = CubicSpline(x, y)
spline_y = spline_interp(new_x)
plt.plot(new_x, spline_y, 'b-', label='三次样条插值')

# 图形美化与显示
plt.legend(loc='lower right', prop=chinese_font)
plt.xlabel('x', fontproperties=chinese_font)
plt.ylabel('y', fontproperties=chinese_font)
plt.title('插值比较: ln(x)', fontproperties=chinese_font)
plt.grid(True)
plt.show()

File: 数据预处理/插值/code1.m
%% 注意：代码文件仅供参考，一定不要直接用于自己的数模论文中
%% 国赛对于论文的查重要求非常严格，代码雷同也算作抄袭
%% 全套课程购买地址：https://k.youshop10.com/NjNu80mX
%% 全套教材PPT请关注微信公众号：大师兄的知识库

x = 1:10;
y = log(x); %lnx
plot(x,y,'o') % 描点作图
hold on;
new_x = 0.01:0.1:10;
p = pchip(x,y,new_x); % 三次埃尔米特插值法
plot(new_x,p)
hold on;
p = spline(x,y,new_x);% 三次样条插值
plot(new_x,p,'b-')
legend('样本点', '三次埃尔米特插值', '三次样条插值', 'Location', 'SouthEast');

%% 注意：代码文件仅供参考，一定不要直接用于自己的数模论文中
%% 国赛对于论文的查重要求非常严格，代码雷同也算作抄袭
%% 全套课程购买地址：https://k.youshop10.com/NjNu80mX
%% 全套教材PPT请关注微信公众号：大师兄的知识库